### PR TITLE
Drop in config

### PR DESCRIPTION
**Pull request by ericcurtin**
For setting configs in a config file

### CODE REVIEW

**Comment by ericcurtin** [comment]
@conan-kudo PTAL , haven't tested or anything yet

**Comment by ericcurtin** [comment]
Tests to fix

File: ramalama/cli.py [hunk 1]
```diff
@@ -52,8 +52,63 @@ def add_argument(self, *args, help=None, default=None, **kwargs):
         super().add_argument(*args, **kwargs)
 
 
+def parse_config_file(file_path, config):
```
Comment (rhatdan) [inline ramalama/cli.py:55]: Lets move the config parsing to a separate file config.py?

Isn't there a standard toml parser written in Python that we can use?

Comment (rhatdan) [reply ramalama/cli.py:55]:
```
import tomllib

with open("/usr/share/containers/containers.conf", "rb") as f:
    data = tomllib.load(f)

print(data)

```

Comment (ericcurtin) [reply ramalama/cli.py:55]: tomllib is available since 3.11, not available on all platforms.

**Comment by Conan-Kudo** [comment]
I'd recommend using `tomlkit` here if you want to use TOML for configuration.

**Comment by ericcurtin** [comment]
The thing is, if we write a simple toml parser like below, which is easy to maintain. We don't necessarily have to worry about whether somebody has that python library available on platform X:

```
import re

class TOMLParser:
    def __init__(self, toml_string=""):
        self.data = {}
        if toml_string:
            self.parse(toml_string)

    def parse(self, toml_string):
        current_section = self.data
        for line in toml_string.splitlines():
            line = line.strip()
            if not line or line.startswith("#"):
                continue

            if line.startswith("[") and line.endswith("]"):
                section_name = line[1:-1].strip()
                current_section = self._create_section(section_name)
            elif "=" in line:
                key, value = line.split("=", 1)
                key = key.strip()
                value = self._parse_value(value.strip())
                current_section[key] = value
            else:
                raise ValueError(f"Invalid TOML line: {line}")
        return self.data

    def _create_section(self, section_name):
        keys = section_name.split(".")
        section = self.data
        for key in keys:
            if key not in section:
                section[key] = {}
            section = section[key]
        return section

    def _parse_value(self, value):
        if value.startswith('"') and value.endswith('"'):
            return value[1:-1]
        elif value.startswith("[") and value.endswith("]"):
            return [self._parse_value(v.strip()) for v in value[1:-1].split(",")]
        elif re.match(r'^\d+$', value):
            return int(value)
        elif re.match(r'^\d+\.\d+$', value):
            return float(value)
        elif value.lower() in {"true", "false"}:
            return value.lower() == "true"
        else:
            raise ValueError(f"Unsupported value type: {value}")

    def get(self, key, default=None):
        keys = key.split(".")
        value = self.data
        for key in keys:
            value = value.get(key)
            if value is None:
                return default
        return value
```

**Comment by ericcurtin** [comment]
tomlkit is only in EPEL which isn't ideal. toml and tomli is in AppStream. But thinking outside of RHEL/Fedora I'd rather just write a small toml parser to cover macOS and other random Linux distros.

File: README.md [hunk 2]
```diff
@@ -15,7 +15,7 @@ Running in containers eliminates the need for users to configure the host system
 
 RamaLama then pulls AI Models from model registries. Starting a chatbot or a rest API service from a simple single command. Models are treated similarly to how Podman and Docker treat container images.
 
-When both Podman and Docker are installed, RamaLama defaults to Podman, The `RAMALAMA_CONTAINER_ENGINE=docker` environment variable can override this behavior. When neather are installed RamaLama will attempt to run the model with software on the local system.
+When both Podman and Docker are installed, RamaLama defaults to Podman, The `RAMALAMA_CONTAINER_ENGINE=docker` environment variable can override this behaviour. When neather are installed RamaLama will attempt to run the model with software on the local system.
```
Comment (ericcurtin) [inline README.md:18]: Today I learned behavior vs behaviour is an US English vs UK English thing. The UK version for whatever reason is more prominent for this word. Can default to US English though if needs be.

**Comment by rhatdan** [comment]
Will your patches work on Windows?

We need to have multi-levels of defaults.

```
Unix Descriptions
            Built-in defaults
            Overwritten by:
                     /usr/share/ramamalam/containers.conf
            Overwritten by:
                     /etc/ramamalam/containers.conf
            Overwritten by:
                     /usr/share/ramamalam/containers.conf.d/*
            Overwritten by:
                     $HOME/.config/ramamalam/containers.conf
            Overwritten by:
                     Environment Variables
            Overwritten by command line options:
                    --engine=foobar
```


            Overwritten by:

**Comment by ericcurtin** [comment]
> Will your patches work on Windows?
>

This will work using WSL2. There's a lot of things in RamaLama that won't work on Windows natively. For one, we use symlinks, which there isn't an exact equivalent for in Windows, although there are similar things.

> We need to have multi-levels of defaults.
> 
> ```
> Unix Descriptions
>             Built-in defaults
>             Overwritten by:
>                      /usr/share/ramamalam/containers.conf
>             Overwritten by:
>                      /etc/ramamalam/containers.conf
>             Overwritten by:
>                      /usr/share/ramamalam/containers.conf.d/*
>             Overwritten by:
>                      $HOME/.config/ramamalam/containers.conf
>             Overwritten by:
>                      Environment Variables
>             Overwritten by command line options:
>                     --engine=foobar
> ```
> 
> ```
>         Overwritten by:
> ```

SGTM

**Comment by ericcurtin** [comment]
Although. This seems funny to me:

```
            Overwritten by:
                     /usr/share/ramamalam/containers.conf
            Overwritten by:
                     /etc/ramamalam/containers.conf
            Overwritten by:
                     /usr/share/ramamala/containers.conf.d/*
            Overwritten by:
                     $HOME/.config/ramamala/containers.conf
            Overwritten by:
                     $HOME/.config/ramamala/containers.conf.d/*
```

I would expect this:

            Overwritten by:
                     /usr/share/ramamala/containers.conf
            Overwritten by:
                     /usr/share/ramamala/containers.conf.d/*
            Overwritten by:
                     /etc/ramamala/containers.conf
            Overwritten by:
                     /etc/ramamala/containers.conf.d/*

**Comment by rhatdan** [comment]
We can leave out .d directories for now, probably overkill for a simple tool.

**Comment by ericcurtin** [comment]
> We can leave out .d directories for now, probably overkill for a simple tool.

I can delete... But it's already implemented, it's just 3 lines of code in the current implementation... Easy to remove though, just delete three lines:

```
        if os.path.exists(path):
            # Load the main config file
            config = parser.parse_file(path)
        if os.path.isdir(path + ".d"):
            # Load all .conf files in ramalama.conf.d directory
            for conf_file in sorted(Path(path + ".d").glob("*.conf")):
                config = parser.parse_file(conf_file)
```

**Pull request merged**